src/phoenix/server/api/dataloaders/evaluation_summaries.py

```diff
@@ -0,0 +1,118 @@
+from collections import defaultdict
+from datetime import datetime
+from typing import (
+    Any,
+    AsyncContextManager,
+    Callable,
+    DefaultDict,
+    List,
+    Literal,
+    Optional,
+    Tuple,
+    Type,
+    Union,
+)
+
+import pandas as pd
+from aioitertools.itertools import groupby
+from sqlalchemy import Select, func, or_, select
+from sqlalchemy.ext.asyncio import AsyncSession
+from strawberry.dataloader import DataLoader
+from typing_extensions import TypeAlias, assert_never
+
+from phoenix.db import models
+from phoenix.server.api.input_types.TimeRange import TimeRange
+from phoenix.server.api.types.EvaluationSummary import EvaluationSummary
+from phoenix.trace.dsl import SpanFilter
+
+Kind: TypeAlias = Literal["span", "trace"]
+ProjectRowId: TypeAlias = int
+TimeInterval: TypeAlias = Tuple[Optional[datetime], Optional[datetime]]
+FilterCondition: TypeAlias = Optional[str]
+EvalName: TypeAlias = str
+
+Segment: TypeAlias = Tuple[Kind, ProjectRowId, TimeInterval, FilterCondition]
+Param: TypeAlias = EvalName
+
+Key: TypeAlias = Tuple[Kind, ProjectRowId, Optional[TimeRange], FilterCondition, EvalName]
+Result: TypeAlias = Optional[EvaluationSummary]
+ResultPosition: TypeAlias = int
+DEFAULT_VALUE = None
+
+
+class EvaluationSummaryDataLoader(DataLoader[Key, Result]):
+    def __init__(self, db: Callable[[], AsyncContextManager[AsyncSession]]) -> None:
+        super().__init__(load_fn=self._load_fn, cache_key_fn=self._cache_key_fn)
+        self._db = db
+
+    @staticmethod
+    def _cache_key_fn(key: Key) -> Tuple[Segment, Param]:
+        kind, project_rowid, time_range, filter_condition, eval_name = key
+        interval = (
+            (time_range.start, time_range.end)
+            if isinstance(time_range, TimeRange)
+            else (None, None)
+        )
+        return (kind, project_rowid, interval, filter_condition), eval_name
+
+    async def _load_fn(self, keys: List[Key]) -> List[Result]:
+        results: List[Result] = [DEFAULT_VALUE] * len(keys)
+        arguments: DefaultDict[
+            Segment,
+            DefaultDict[Param, List[ResultPosition]],
+        ] = defaultdict(lambda: defaultdict(list))
+        for position, key in enumerate(keys):
+            segment, param = self._cache_key_fn(key)
+            arguments[segment][param].append(position)
+        for segment, params in arguments.items():
+            stmt = _get_stmt(segment, *params.keys())
+            async with self._db() as session:
+                data = await session.stream(stmt)
+                async for eval_name, group in groupby(data, lambda row: row.name):
+                    summary = EvaluationSummary(pd.DataFrame(group))
+                    for position in params[eval_name]:
+                        results[position] = summary
+        return results
+
+
+def _get_stmt(
```

Comment:
Another one where i think it would help to break out into separate functions for spans and traces. Makes it harder to see what query is being built.